packages/js/src/executors/tsc/tsc.impl.ts

```diff
@@ -141,13 +159,203 @@ export async function* tscExecutor(
+    context,
+    options
+  );
+
+  const inlinedDependencies = handleInliningBuild(
+    context,
+    options,
+    tsCompilationOptions
+  );
+
   return yield* compileTypeScriptFiles(
     options,
-    createTypeScriptCompilationOptions(options, context),
+    tsCompilationOptions,
     async () => {
       await assetHandler.processAllAssetsOnce();
       updatePackageJson(options, context, target, dependencies);
+
+      if (inlinedDependencies.length > 0) {
+        postProcessInlinedDependencies(
+          tsCompilationOptions,
+          inlinedDependencies
+        );
+      }
+    }
+  );
+}
+
+function handleInliningBuild(
+  context: ExecutorContext,
+  options: NormalizedExecutorOptions,
+  tsCompilationOptions: TypeScriptCompilationOptions
+): Array<ProjectGraphProjectNode & { pathAlias: string }> {
+  const projectDependencies =
+    context.projectGraph.dependencies[context.projectName];
+  const dependencies: Array<{ name: string; path: string }> = [];
+
+  if (projectDependencies && projectDependencies.length) {
+    const tsConfigJson = readJsonFile(tsCompilationOptions.tsConfig);
+
+    // handle inlining
+    // 1. figure out a list of dependencies that need to be inlined
+
+    for (const projectDependency of projectDependencies) {
+      let shouldInline = false;
+
+      if (options.external === 'all') {
+        /**
+         * if all buildable libraries are marked as external,
+         * then push the project dependency that doesn't have a build target
+         */
+        if (!hasBuildTarget(projectDependency.target, context, options)) {
+          shouldInline = true;
+        }
+      } else if (options.external === 'none') {
+        /**
+         * if all buildable libraries are marked as internal,
+         * then push every project dependency to be inlined
+         */
+        shouldInline = true;
+      } else if (options.external.length > 0) {
+        /**
+         * if some buildable libraries are marked as internal,
+         * then push those that do not have a build target AND those with build target but NOT part of options.external
+         */
+        if (
+          !hasBuildTarget(projectDependency.target, context, options) ||
+          !options.external.includes(projectDependency.target)
+        ) {
+          shouldInline = true;
+        }
+      }
+
+      if (shouldInline) {
+        dependencies.push({
+          name: projectDependency.target,
+          path: getPathAliasForPackage(
+            context.projectGraph.nodes[projectDependency.target],
+            tsConfigJson['compilerOptions']['paths']
+          ),
+        });
+      }
+    }
+
+    // 2. adjust rootDir of tmp tsConfig if we have dependencies to inline
+    if (dependencies.length > 0) {
+      logger.log(
+        'List of dependencies to be inlined:',
+        JSON.stringify(dependencies, null, 2)
+      );
+
+      // update rootDir to inline
+      tsCompilationOptions.rootDir = '.';
```

Comment:
Can we set this in `tscExecutor` since that is where `tsCompilationOptions` is created and passed around. It's not clear that calling `handleInliningBuild` would mutate this.

Also, maybe `rootDir: '.'` can be set as such always without checking deps length? Not sure if there is a practical difference here so I could be wrong.